### PR TITLE
fix: Support Oracle blob data containing non-utf8 content

### PR DESCRIPTION
**Pull request by johncrn**
### Feature or Bugfix
- Bugfix

### Detail
- Support Oracle BLOB data containing non-utf8 content
- Currently the `_databases.py` file's `_records2df` function fails when 
  - The dtype for a BLOB column has been set to `string` and the Oracle LOB value returns bytes
  - The dtype for a BLOB column has been set to `binary`
- Note that the `oracle.py` file's `handle_oracle_objects` function expects the `.read()` method of the Oracle results to return a `str`. However, this method may return either a `str` or `bytes` object. Implementation [here](https://github.com/oracle/python-oracledb/blob/999fd2b4695cae51f2063ae1ba4f63750d050c32/src/oracledb/lob.py#L101).
- Note that if the dtype provided for a blob column is set to a pyarrow binary type, the `handle_oracle_objects` function is not called.
- The failure does not always occur if the BLOB column contents contain utf8 compatible data, as in this case Oracle results `.read()` method returns a string

### Relates
- [Related bug ticket](https://github.com/aws/aws-sdk-pandas/issues/2108)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: b5a3860a63d057e7f7172bf818ec29fa50c800c1
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2f40899d-8de5-49f8-88f5-d9133c5c3ef0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 8a3523005677876e2540d2689a97fda23507cd87
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=435a55ec-1d79-4b85-a358-f40d1b4b67d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: bf3d0eb8f7cbd5222b852b4e3f0ec2453e5ffdb0
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=57086687-dbd5-4c3d-8e07-c509b05671da%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 535ea59024b471401ab433f25dd900b920f3caa9
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a6ba2f6d-3eb8-4032-94db-85660f3bf1c0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 81c42cb24523a61ce1b2b36a6297cd3126fc1c6c
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6c2f668c-6c5b-4ca6-81a6-52e242aa115a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 4e1d4ef3e7e8a8f8df871d11506881856f28298a
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=214edcb0-925f-483e-a0cd-1c62705aaf84%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: b1b0ded47a9fe1b652a6a9397a0edf8d75b3d26d
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9f938df9-15c5-4e46-b46b-328326a9b866%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/oracle.py [hunk 1]
```diff
@@ -457,4 +458,13 @@ def handle_oracle_objects(
                 Decimal(repr(col_value)) if isinstance(col_value, float) else col_value for col_value in col_values
             ]
 
+        # A user may wish to represent Oracle blob binary data as plain text - if so base64 encode bytes
```
Comment (cnfait) [inline awswrangler/oracle.py:461]: This sounds like a feature - is that a fairly common use case? while the overall PR looks good (thanks a lot!) I'm not sure about this specific behaviour...

Comment (johncrn) [reply awswrangler/oracle.py:461]: > This sounds like a feature - is that a fairly common use case? while the overall PR looks good (thanks a lot!) I'm not sure about this specific behaviour...

Hi @cnfait ,

Thanks for taking a look! Yes that is true, in essence representing binary as a string is a form of transformation which can be viewed as a new feature/behaviour. Happy to drop it if that sits better with you, simply remove that block. 

Also my apologies for not being able to get the final check passing. I've not been able to stand up a full dev environment to run through the tests.

Comment (johncrn) [reply awswrangler/oracle.py]: Hi @cnfait , I've pushed a commit which reverts this behaviour.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: b22d6d7bef9d94483a41d3ff09bf85af40e9ced2
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e3cc7e22-3b50-4be9-9e3d-4fcc066b1fba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/oracle.py (hunk 1 above)
Comment (cnfait) [reply awswrangler/oracle.py]: Thanks again!

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: b22d6d7bef9d94483a41d3ff09bf85af40e9ced2
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7a1ccfaa-3cc4-4888-b226-beb1e8964e03%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 4dcd7a5b8316caecdc50be36daf6cf732faf770f
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3cb500d6-11a5-4760-9fd6-0dfac7022e43%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by kukushking — APPROVED** [review]
Thank you for the contribution!

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: b8add407280dd422b76907c26acb220219122f65
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9dc16b8d-5972-4653-aa45-0c084afb1250%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 5257d04e50a83391c330af10ec33c43808855221
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2bdc28ba-4023-4364-a41b-de872d79811a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by kukushking** [comment]
Hi @johncrn @cnfait there is still one related test failing `tests/test_oracle.py::test_sql_types`:
```
    def test_sql_types(oracle_table, oracle_con):
        table = oracle_table
        df = get_df()
>       wr.oracle.to_sql(
            df=df,
            con=oracle_con,
            table=table,
            schema="TEST",
            mode="overwrite",
            index=True,
            dtype={"iint32": "NUMBER(10)", "decimal": "NUMBER(3,2)"},
        )

tests/test_oracle.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
awswrangler/oracle.py:35: in inner
    return func(*args, **kwargs)
awswrangler/_config.py:542: in wrapper
    return function(**args)
awswrangler/oracle.py:397: in to_sql
    _create_table(
awswrangler/oracle.py:106: in _create_table
    cursor.execute(sql)
.tox/py38/lib/python3.8/site-packages/oracledb/cursor.py:378: in execute
    impl.execute(self)
src/oracledb/impl/thin/cursor.pyx:121: in oracledb.thin_impl.ThinCursorImpl.execute
    ???
src/oracledb/impl/thin/protocol.pyx:375: in oracledb.thin_impl.Protocol._process_single_message
    ???
src/oracledb/impl/thin/protocol.pyx:376: in oracledb.thin_impl.Protocol._process_single_message
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   oracledb.exceptions.DatabaseError: ORA-00906: missing left parenthesis
```

**Comment by cnfait** [comment]
@kukushking I'm looking into it

**Comment by johncrn** [comment]
Thanks @cnfait , and apologies I wasn't able to iron this one out myself.

I believe this error started occurring because I removed this line from the test suite: https://github.com/aws/aws-sdk-pandas/pull/2107/files#diff-3f3c8e171ca0635eaadb40bf6ea260d38d735529b7bedde4c3ac5f8bd07069e1L41

When I spotted that line I suspected it may have been added to work around the fact Oracle blob types were not supported for data ingestion, so by removing it this fix should be able to be validated. But perhaps there are other underlying issues at play.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 389c60d8e1dcd119fa0846903b4403eee5886728
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=683af9d0-05f8-466d-a051-e30d94fbf2c0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by johncrn** [comment]
Hi @cnfait ,

Apologies I missed your PR before making my earlier comment. I've merged in that change but the codebuild job has failed. Had a quick look at the logs and wonder if the error was transient?

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: e48e5f26f9a05e1e85ef9f27609f56c52e70e4f4
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0504f2b8-f6c5-4def-acf4-4f20732348ae%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**